### PR TITLE
batches: add more restrictions to viewing batch changes via URL and executions

### LINKED ISSUE #40157: batches: non-site-admin cannot view drafts, execution status in main BC list
**EDIT** See comment below for updated findings.

**Steps to repro:**

1. Create a batch change with server-side UI but leave it in the draft state
2. Create a batch change and execute the spec with server-side UI, but don't apply it
3. View the batch changes list page as a site admin
4. View the batch changes list page as a non-site admin

**Expected behavior:**
- The draft BC is visible, and the BC with the active execution has the execution status pill inset

**Actual behavior:**
- Only the site admin can see the draft BC and execution status pill inset

### PR DESCRIPTION
**Pull request by adeola-ak**
closes #40157

Previously everyone could see batch specs if they had the ID but this fix is to viewing make draft batch changes consistent with the behavior of the batch change list (hidden) via URL as well as remove the executions tab when viewing another users batch change as a non admin or non same user

## Test plan

shared executions url to a non admin who was met with a 404 error. Also not able to read data from graphql console

### CODE REVIEW

**Comment by BolajiOlajide** [comment]
@adeola-ak I see you requested review here but it's still a draft. Can you clarify if this is ready for review or I should hold on?

**Comment by adeola-ak** [comment]
@BolajiOlajide I do think there was one more thing i wanted to do on this but now im not sure, you can review now! Thanks

**Review by BolajiOlajide — APPROVED** [review]
CleanShot 2022-10-04 at 16 15 08@2x

Do we want to remove the `Error:` prefix here in the messaging? I'm not sure where it comes from.
cc @danielmarquespt

File: enterprise/cmd/frontend/internal/batches/resolvers/resolver.go [hunk 1]
```diff
@@ -264,8 +263,17 @@ func (r *Resolver) batchSpecByID(ctx context.Context, id graphql.ID) (graphqlbac
 		return nil, err
 	}
 
-	// Everyone can see batch specs, if they have the ID.
-	return &batchSpecResolver{store: r.store, batchSpec: batchSpec}, nil
+	specResolver := &batchSpecResolver{store: r.store, batchSpec: batchSpec}
+	canAdminister, err := specResolver.ViewerCanAdminister(ctx)
```
Comment (Piszmog) [inline enterprise/cmd/frontend/internal/batches/resolvers/resolver.go:267]: Please add a test for the scenario where a user does not have permissions tries to view the batch spec

**Comment by courier-new** [comment]
No strong feelings on this, but are we okay with the fact that this still reveals the existence of a batch spec with this ID? In theory it's not like a malicious actor would be able to infer any additional information from that, but they could still plug the ID in on other places in the API and _try_ to. Generally if a user doesn't have permission to access something potentially sensitive, I might just default to a 404. But then you sacrifice the UX of the nicer error message, of course.

**Comment by chrispine** [comment]
Great question, Kelli! Personally, I'd prefer the better error message and am okay with this, but I also don't feel strongly and am open to other opinions.

**Comment by eseliger** [comment]
To be consistent with how we handle those cases elsewhere, I would probably default to the 404 behavior, but no terribly strong feelings

**Comment by adeola-ak** [comment]
i updated to 404 page to be consistent. should i add somewhere in our docs that this behavior is not permitted, being that i'm. returning a 404 error without any further detail of why i would be getting it as a user?

**Comment by courier-new** [comment]
That's a great idea. We don't really talk about batch spec execution permissions in our _doc_ docs, but I totally think this makes sense to include on the GraphQL schema docs. Maybe you could add a note to the docstring for the `BatchSpec` type, like "For security, a batch spec is only visible to viewers who can administer it."

**Pull request merged**